homeassistant/components/tradfri/base_class.py

```diff
@@ -44,102 +35,31 @@ async def wrapper(command: Command | list[Command]) -> None:
     ) -> None:
         """Initialize a device."""
-        self._api = handle_error(api)
-        self._attr_name = device.name
-        self._device: Device = device
-        self._device_control: BlindControl | LightControl | SocketControl | SignalRepeaterControl | AirPurifierControl | None = (
-            None
-        )
-        self._device_data: Socket | Light | Blind | AirPurifier | None = None
-        self._gateway_id = gateway_id
-
-    async def _async_run_observe(self, cmd: Command) -> None:
-        """Run observe in a coroutine."""
-        try:
-            await self._api(cmd)
-        except PytradfriError as err:
-            self._attr_available = False
-            self.async_write_ha_state()
-            _LOGGER.warning("Observation failed, trying again", exc_info=err)
-            self._async_start_observe()
-
-    @callback
-    def _async_start_observe(self, exc: Exception | None = None) -> None:
-        """Start observation of device."""
-        if exc:
-            self._attr_available = False
-            self.async_write_ha_state()
-            _LOGGER.warning("Observation failed for %s", self._attr_name, exc_info=exc)
-        cmd = self._device.observe(
-            callback=self._observe_update,
-            err_callback=self._async_start_observe,
-            duration=0,
-        )
-        self.hass.async_create_task(self._async_run_observe(cmd))
-
-    async def async_added_to_hass(self) -> None:
-        """Start thread when added to hass."""
-        self._async_start_observe()
-
-    @callback
-    def _observe_update(self, device: Device) -> None:
-        """Receive new state data for this device."""
-        self._refresh(device)
+        super().__init__(device_coordinator)
 
-    def _refresh(self, device: Device, write_ha: bool = True) -> None:
-        """Refresh the device data."""
-        self._device = device
-        self._attr_name = device.name
-        if write_ha:
-            self.async_write_ha_state()
-
-
-class TradfriBaseDevice(TradfriBaseClass):
-    """Base class for a TRADFRI device.
+        self._gateway_id = gateway_id
 
-    All devices should inherit from this class.
-    """
+        self._device: Device = device_coordinator.data
+        self._attr_available = self._device.reachable
 
-    def __init__(
-        self,
-        device: Device,
-        api: Callable[[Command | list[Command]], Any],
-        gateway_id: str,
-    ) -> None:
-        """Initialize a device."""
-        self._attr_available = device.reachable
-        self._hub_available = True
-        super().__init__(device, api, gateway_id)
-
-    async def async_added_to_hass(self) -> None:
-        """Start thread when added to hass."""
-        # Only devices shall receive SIGNAL_GW
-        self.async_on_remove(
-            async_dispatcher_connect(self.hass, SIGNAL_GW, self.set_hub_available)
-        )
-        await super().async_added_to_hass()
+        self._coordinator = device_coordinator
```

Comment:
The coordinator can be accessed as `self.coordinator` in the coordinator entity.
```suggestion
```